File: 0011-container-with-most-water/0011-container-with-most-water.py
"""
Time = O(n)
Space = O(1)
"""
class Solution:
    def maxArea(self, height: List[int]) -> int:
        # Create two pointers
        left = 0
        right = len(height) - 1
        cur_max_area = 0
        
        # Traverse through the array
        while left < right:
            # Calculate the current area
            w = right - left
            h = min(height[left], height[right])
            cur_max_area = max(cur_max_area, w * h)
            # move one of the pointer which points to the shorter wall
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
                
        return cur_max_area

File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0, right = height.size() - 1;
        int left_h, right_h;
        int maxVol = 0;

        int w, h;
        while (left < right) {
            left_h = height[left];
            right_h = height[right];

            w = right - left;
            h = min(left_h, right_h);
            maxVol = max(maxVol, w * h);

            if (left_h < right_h) {
                left++;
            } else {
                right--;
            }
        }

        return maxVol;
    }
};